staging/src/k8s.io/apiserver/pkg/util/proxy/streamtranslator.go

```diff
@@ -0,0 +1,149 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package proxy
+
+import (
+	"fmt"
+	"net/http"
+	"net/url"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/httpstream/spdy"
+	utilnet "k8s.io/apimachinery/pkg/util/net"
+	apimachineryproxy "k8s.io/apimachinery/pkg/util/proxy"
+	constants "k8s.io/apimachinery/pkg/util/remotecommand"
+	"k8s.io/client-go/tools/remotecommand"
+	"k8s.io/client-go/util/exec"
+	"k8s.io/klog/v2"
+)
+
+// StreamTranslatorHandler is a handler which translates WebSocket stream data
+// to SPDY to proxy to kubelet (and ContainerRuntime).
+type StreamTranslatorHandler struct {
+	// Location is the location of the upstream proxy. It is used as the location to Dial on the upstream server
+	// for upgrade requests.
+	Location *url.URL
+	// Transport provides an optional round tripper to use to proxy. If nil, the default proxy transport is used
+	Transport http.RoundTripper
+	// Responder is passed errors that occur while setting up proxying.
+	Responder apimachineryproxy.ErrorResponder
+	// Options define the requested streams (e.g. stdin, stdout).
+	Options Options
+}
+
+// NewStreamTranslatorHandler creates a new proxy handler. Responder is required for returning
+// errors to the caller.
+func NewStreamTranslatorHandler(location *url.URL, transport http.RoundTripper, responder apimachineryproxy.ErrorResponder, opts Options) *StreamTranslatorHandler {
+	return &StreamTranslatorHandler{
+		Location:  location,
+		Transport: transport,
+		Responder: responder,
+		Options:   opts,
+	}
+}
+
+func (h *StreamTranslatorHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	// Create WebSocket server, including particular streams requested.
+	websocketStreams, err := webSocketServerStreams(req, w, h.Options)
+	if err != nil {
+		klog.Errorf("websocket createStreams error: %v", err)
+		h.Responder.Error(w, req, err)
+		return
+	}
+	defer websocketStreams.conn.Close()
+
+	// Create the SPDY client connection proxied through kubelet.
+	tlsConfig, err := utilnet.TLSClientConfig(h.Transport)
```

Comment:
extracting and reusing things piecemill from this transport isn't correct... this extracts the tlsConfig but drops any custom Dialer/Proxier from the transport - https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/client/kubelet_client.go/#L99-L108

this means the constructed spdy client won't flow through the konnectivity network proxy correctly but will try to dial directly

Reply:
🤕